### PR TITLE
Use ZeroMemset in View constructor

### PR DESCRIPTION
**Pull request by masterleinad**
Follow-up to https://github.com/kokkos/kokkos/pull/3944.
This pull request makes sure that we also use `std::memset` and similar in the `View` construction.
Since `ViewValueFunctor::execute` is not a `KOKKOS_FUNCTION` it doesn't make sense to ask for `space.in_parallel()` IMHO, so I also removed that branch. I am happy to discuss that separately, of course.

### CODE REVIEW

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 1]
```diff
@@ -2817,6 +2817,22 @@ struct ViewDataHandle<
 namespace Kokkos {
 namespace Impl {
 
+template <typename T>
```
Comment (DavidPoliakoff) [inline core/src/impl/Kokkos_ViewMapping.hpp:2820]: Why was this moved?

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 2]
```diff
@@ -2867,33 +2883,43 @@ struct ViewValueFunctor<ExecSpace, ValueType, false /* is_scalar */> {
...
+          space,
+          Kokkos::View<ValueType*, typename DeviceType::memory_space,
+                       Kokkos::MemoryTraits<Kokkos::Unmanaged>>(ptr, n),
+          value);
+      return;
+    }
+
     PolicyType policy(0, n);
     std::string functor_name;
-    if (!space.in_parallel()) {
```
Comment (DavidPoliakoff) [inline core/src/impl/Kokkos_ViewMapping.hpp:2872]: This is for use cases involving Views of Views, we need that branch, I believe

**Review by DavidPoliakoff — CHANGES_REQUESTED** [review]
I'm requesting changes w.r.t. the `in_parallel` question, the other is just a comment. Note that this is just a first pass, I'll look for other stuff in detail

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_ViewMapping.hpp:2820]: Moving it here compiles, i.e. the include order works this way, leaving it in the original place does not.
We could of course define it in a separate header file instead.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_ViewMapping.hpp:2872]: But then it should still be a `KOKKOS_FUNCTION`, doesn't it? Anyway, I will just separate it into a different pull request.

Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_ViewMapping.hpp:2872]: Sure. For reference, it isn't KOKKOS_FUNCTION, because the _outer_ constructor still happens on the Host. It's really confusing, everything about View of Views is weird

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 1 above)
Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_ViewMapping.hpp:2820]: This is fine by me, just curious

**Review by DavidPoliakoff — APPROVED** [review]

**Review by crtrott — APPROVED** [review]

File: core/unit_test/CMakeLists.txt [hunk 3]
```diff
@@ -783,7 +783,7 @@ KOKKOS_ADD_ADVANCED_TEST( UnitTest_PushFinalizeHook_terminate
...
+      PASS_REGULAR_EXPRESSION "kokkosp_init_library::kokkosp_parse_args:4:KokkosCore_ProfilingAllCalls:-c:test:delimit::.*::kokkosp_allocate_data:${MEMSPACE_REGEX}:source:${ADDRESS_REGEX}:40::kokkosp_allocate_data:${MEMSPACE_REGEX}:destination:${ADDRESS_REGEX}:40::kokkosp_begin_deep_copy:${MEMSPACE_REGEX}:destination:${ADDRESS_REGEX}:${MEMSPACE_REGEX}:source:${ADDRESS_REGEX}:40::kokkosp_end_deep_copy::kokkosp_begin_parallel_for:parallel_for:${SIZE_REGEX}:0::kokkosp_end_parallel_for:0::kokkosp_begin_parallel_reduce:parallel_reduce:${SIZE_REGEX}:1.*::kokkosp_end_parallel_reduce:1::kokkosp_begin_parallel_scan:parallel_scan:${SIZE_REGEX}:2::kokkosp_end_parallel_scan:2::kokkosp_push_profile_region:push_region::kokkosp_pop_profile_region::kokkosp_create_profile_section:created_section:3::kokkosp_start_profile_section:3::kokkosp_stop_profile_section:3::kokkosp_destroy_profile_section:3::kokkosp_profile_event:profiling_event::kokkosp_declare_metadata:dogs:good::kokkosp_deallocate_data:${MEMSPACE_REGEX}:destination:${ADDRESS_REGEX}:40::kokkosp_deallocate_data:${MEMSPACE_REGEX}:source:${ADDRESS_REGEX}:40::kokkosp_finalize_library::"
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt:786]: @DavidPoliakoff @crtrott please confirm you looked at these changes carefully

Comment (masterleinad) [reply core/unit_test/CMakeLists.txt:786]: What happens here is that don't run a `Kokkos::parallel_for` anymore (but `ZeroMemset`) and thus also don't print related profiling information.

Comment (dalg24) [reply core/unit_test/CMakeLists.txt:786]: This means we lose the ability to detect unnecessary view initialization with the tools

Comment (masterleinad) [reply core/unit_test/CMakeLists.txt:786]: Yes. Would you prefer us slightly lying by keeping the profiling information in place?
